File: core/src/main/scala/com/github/propi/rdfrules/ruleset/RulesetWriter.scala
package com.github.propi.rdfrules.ruleset

import java.io.File
import com.github.propi.rdfrules.data.Compression
import com.github.propi.rdfrules.data.formats.Compressed
import com.github.propi.rdfrules.rule.ResolvedRule
import com.github.propi.rdfrules.utils.{ForEach, OutputStreamBuilder}

import scala.util.Try

/**
  * Created by Vaclav Zeman on 18. 4. 2018.
  */
trait RulesetWriter {
  def writeToOutputStream(rules: ForEach[ResolvedRule], outputStreamBuilder: OutputStreamBuilder): Unit

  def writeToOutputStream(ruleset: Ruleset, outputStreamBuilder: OutputStreamBuilder): Unit = writeToOutputStream(ruleset.resolvedRules, outputStreamBuilder)
}

object RulesetWriter extends Compressed {

  implicit object NoWriter extends RulesetWriter {
    def writeToOutputStream(rules: ForEach[ResolvedRule], outputStreamBuilder: OutputStreamBuilder): Unit = throw new IllegalStateException("No specified RulesetWriter.")
  }

  def apply(file: File): RulesetWriter = {
    val Ext2 = ".+[.](.+)[.](.+)".r
    val Ext1 = ".+[.](.+)".r
    file.getName.toLowerCase match {
      case Ext2(ext1, ext2) => Try(Compression(ext2)).toOption match {
        case Some(compression) => RulesetSource(ext1).compressedBy(compression)
        case None => RulesetSource(ext2)
      }
      case Ext1(ext1) => RulesetSource(ext1)
      case _ => throw new IllegalArgumentException("No file extension to detect a Ruleset format.")
    }
  }

}

File: java/src/main/java/com/github/propi/rdfrules/java/ruleset/RulesetWriter.java
package com.github.propi.rdfrules.java.ruleset;

import com.github.propi.rdfrules.java.ReadersWriters;
import com.github.propi.rdfrules.java.ScalaConverters;
import com.github.propi.rdfrules.java.rule.ResolvedRule;

import java.io.OutputStream;
import java.util.function.Supplier;

/**
 * Created by Vaclav Zeman on 13. 5. 2018.
 */
public interface RulesetWriter {

    void writeToOutputStream(Iterable<ResolvedRule> rules, Supplier<OutputStream> osb);

    class Json implements RulesetWriter {
        @Override
        public void writeToOutputStream(Iterable<ResolvedRule> rules, Supplier<OutputStream> osb) {
            ReadersWriters.rulesJsonWriter().writeToOutputStream(
                    ScalaConverters.toIterable(rules, ResolvedRule::asScala),
                    osb::get
            );
        }
    }

    class Text implements RulesetWriter {
        @Override
        public void writeToOutputStream(Iterable<ResolvedRule> rules, Supplier<OutputStream> osb) {
            ReadersWriters.rulesTextWriter().writeToOutputStream(
                    ScalaConverters.toIterable(rules, ResolvedRule::asScala),
                    osb::get
            );
        }
    }

}